neps/nep-0541.md

```diff
@@ -0,0 +1,196 @@
+Receipts that spawn from a transaction have the same priority as the transaction itself.
+
+Regardless of priority fee, a transaction still needs to pay the base cost, which is determined by `minimum_gas_price`. The current dynamic gas price mechanism will be removed since it does not make sense for the congestion of one shard to affect the gas price of transaction in a completely different shard, even if there is no congestion there.
+
+When a chunk is produced, the transactions should be sorted by priority in descending order. Otherwise the chunk is invalid.
+However, when the priority of two transactions are the same, no specific order is enforced by the protocol.
+
+When a transaction is processed and converted to a receipt, X percent (X is TBD) of the priority fee is burnt and the rest is rewarded to the chunk producer.
+During receipt processing, receipts with highest priorites (from incoming and delayed receipts combined) will get processed first, but up to a fixed amount of gas (the threshold is TBD). The rest of the gas limit is devoted to processing all receipts, regardless of priority, in a FIFO order. This is to ensure that the system does not deadlock in light of [NEP-539](https://github.com/near/NEPs/pull/539).
+
+## Reference Implementation
+
+We discuss the implementation of two important pieces of the proposal, receipt processing and validator reward, in this section.
+
+### Receipt Processing
+
+A max heap is created in the state for storing priorities of delayed receipts. More specifically, we add a trie column `RECEIPTS_PRIORITY_QUEUE: u8 = 13`. Each value in this heap is of the form
+```rust
+struct ReceiptPriority {
+    /// index of the corresponding receipt in the delayed receipts queue. Use u32 here to save some space in the state
+    receipt_index: u32
+    /// priority of the receipt
+    priority: u64
+}
+```
+Since each value is quite small and we don't expect the delayed receipt queue to be extremely long (even if there are 100k delayed receipts, it is only 1.2MB in size), and it only needs to be read and write once per chunk, this heap can be stored as a single value in the state.
+
+The receipt processing order is roughly as follows:
+
+```python
+# Delayed receipts is a max heap and incoming receipts is sorted by priority in descending order
+# gas limit refers to the gas limit assigned to process receipts with priority
+def process_priority_receipts(delayed_receipts, incoming_receipts, gas_limit):
+    if delayed_receipts.empty() and incoming_receipts.empty():
+        return
+    gas_used = 0
+    while gas_used < gas_limit:
+        delayed_receipt_head = if delayed_receipts.empty() {-Inf} else {delayed_receipts.top() }
+        incoming_receipt_head = if incoming_receipts.empty() {-Inf} else {incoming_receipts.top() }
+        receipt = None
+        if delayed_receipt_head.priority > incoming_receipts_head.priority: 
```

Comment:
Are prioritized delayed receipts executed before local receipts? What about prioritized local receipts?
Today's order among receipts is 
- local receipts
- delayed receipts
- new incoming receipts

Note that if we don't execute local receipts in the first chunk, they end up in the delayed receipts queue one chunk later.

Reply:
yes they are executed before local receipts. The way to think of it is as follows: priority execution always happens first and regular execution happens. During regular execution, we preserve the order of execution we do today (local receipts, delayed receipts, and incoming receipts). We can also change the order within regular execution, but that is orthogonal to this proposal.